### PR TITLE
f/design of experiments

### PR DESCRIPTION
**Pull request by nikhar-abbas**
# Design of experiments
This PR adds the capability to run design of experiments testing to better understand the optimization space. This enables us to leverage OpenMDAO to study the design space of the design variables with respect to our numerous merit figures. 

## Major changes:
Broke up the `driver` section in analysis_options.yaml to include `optimization` and `design_of_experiments` subsections. These also now include flags to distinguish which driver type you would like to run. This comes with changes to gc_PoseOptimization.py to account for this. 

## Notes on the implementation
1. Only uniform model generation is implemented, currently. These can easily be expanded to all of those offered [through OpenMDAO](http://openmdao.org/twodocs/versions/latest/features/building_blocks/drivers/doe_driver.html), it just hasn't been done yet
2. The bug fix in #60 is not included in this PR, though it should be. After incorporating the fixes in #60 locally, I found some issues with Frame3DD being happy if I tried to look at TSR as a design variable, this could be user error though.
3. The restructure in #62 certainly puts a wrinkle in all of this. I figured I would open this as a draft PR before trying to restructure things within the context of #62, though I don't think it should be too tricky.

### CODE REVIEW

**Comment by gbarter** [comment]
Nikhar- I really like the changes you are proposing.  I actually just submitted my own PR to make the gc_PoseOpt files between WEIS and WISDEM play nicely together.  I would actually advocate for most of these code changes being made on the WISDEM side so that both codes can benefit.  I think we might have to edit the analysis options schema too.  I would be happy to do this if you like.

Also want to tag @ptrbortolotti

**Comment by nikhar-abbas** [comment]
> Nikhar- I really like the changes you are proposing. I actually just submitted my own PR to make the gc_PoseOpt files between WEIS and WISDEM play nicely together. I would actually advocate for most of these code changes being made on the WISDEM side so that both codes can benefit. I think we might have to edit the analysis options schema too. I would be happy to do this if you like.
> 
> Also want to tag @ptrbortolotti

Garrett, this makes a lot of sense. I can take a crack at getting a PR into WISDEM with these updates so that your changes in #62 account for these changes. 

I would like to do some more extensive testing, especially with OpenFAST, before totally marking this PR as ready also. That'll involve more WEIS specific stuff, I suppose.

**Comment by nikhar-abbas** [comment]
Okay - I have made the changes to WISDEM and submitted a PR [here](https://github.com/WISDEM/WISDEM/pull/216).

Once those are approved, I will update the subtree in WEIS and change this PR accordingly.

**Comment by gbarter** [comment]
That was quick!  I will take a look at the WISDEM PR and may contribute to it.

**Comment by johnjasa** [comment]
Hey @nikhar-abbas, I'm looping back to this. There were some versioning issues with the CI packages, but I pushed a hotfix and now all tests are passing. Were there any further changes you were thinking for this, or is it ready to move from draft state to mergeable?

**Comment by nikhar-abbas** [comment]
> Hey @nikhar-abbas, I'm looping back to this. There were some versioning issues with the CI packages, but I pushed a hotfix and now all tests are passing. Were there any further changes you were thinking for this, or is it ready to move from draft state to mergeable?

@johnjasa glad the CI got sorted! I'm still fighting through some MPI stuff that will involve some more commits - will let you know when I think it is ready for a merge.

**Comment by nikhar-abbas** [comment]
Hey @johnjasa,

This seems to be working somewhat well for me, but definitely warrants a somewhat thorough review before being merged (even if the CI passes). Once you think its good we can mark mark it ready for review

**Comment by johnjasa** [comment]
Great, @nikhar-abbas, I'll start checking it out! Could you resolve the conflicting files here by pulling from the latest develop, merging, and pushing back up?

**Comment by nikhar-abbas** [comment]
> Great, @nikhar-abbas, I'll start checking it out! Could you resolve the conflicting files here by pulling from the latest develop, merging, and pushing back up?

Thought I did that but I guess I messed something up! All fixed and CI is running now.

File: weis/glue_code/runWEIS.py [hunk 1]
```diff
@@ -115,6 +115,9 @@ def run_weis(fname_wt_input, fname_modeling_options, fname_opt_options, overridd
             wt_opt = myopt.set_design_variables(wt_opt, wt_init)
             wt_opt = myopt.set_constraints(wt_opt)
             wt_opt = myopt.set_recorders(wt_opt)
+
+            if opt_options['driver']['design_of_experiments']:
```
Comment (ptrbortolotti) [inline weis/glue_code/runWEIS.py:119]: @nikhar-abbas shouldn't this be 

if opt_options['driver']['design_of_experiments']['flag']:

Comment (nikhar-abbas) [reply weis/glue_code/runWEIS.py]: Yes, fixed!

File: docs/conf.py [hunk 2]
```diff
@@ -58,6 +58,32 @@
 # The master toctree document.
 master_doc = 'index'
 
+# General information about the project.
```
Comment (johnjasa) [inline docs/conf.py:61]: It looks like this is the info for OpenFAST, but within the WEIS conf.py. Is this on purpose?

File: docs/requirements.txt [hunk 3]
```diff
@@ -1,17 +1,9 @@
...
-
-# docs
-sphinx>2.0
-numpydoc
-numpydoc
-sphinxcontrib-bibtex<2.0.0
-sphinx-rtd-theme
-# sphinx-autoapi
\ No newline at end of file
+breathe>=4.4
```
Comment (johnjasa) [inline docs/requirements.txt:1]: Are these changes purposeful too? It looks like they don't include a few that are required for WEIS -- like OpenMDAO.

File: examples/03_blade/analysis_options_aero.yaml [hunk 4]
```diff
@@ -0,0 +1,235 @@
+general:
```
Comment (johnjasa) [inline examples/03_blade/analysis_options_aero.yaml:1]: It looks like these three files come from the WISDEM `examples` folder and the `03` designation would overlap with the existing WEIS `03` example.

File: examples/05_tower_monopile/analysis_options.yaml [hunk 5]
```diff
@@ -0,0 +1,62 @@
+general:
```
Comment (johnjasa) [inline examples/05_tower_monopile/analysis_options.yaml:1]: This appears to come from the WISDEM examples again, like the previous 03 example clash

File: examples/13_design_of_experiments/IEA-15-240-RWT.yaml [hunk 6]
```diff
@@ -0,0 +1,970 @@
+name: IEA 15MW Offshore Reference Turbine, with taped chord tip design
```
Comment (johnjasa) [inline examples/13_design_of_experiments/IEA-15-240-RWT.yaml:1]: I think these files are duplicated from WISDEM as well

File: modules/moordyn/src/MoorDyn_Driver.f90 [hunk 7]
```diff
@@ -0,0 +1,416 @@
+!**********************************************************************************************************************************
```
Comment (johnjasa) [inline modules/moordyn/src/MoorDyn_Driver.f90:1]: I think this is here by accident

**Review by johnjasa — CHANGES_REQUESTED** [review]
Looks great! I'm going to stress-test with some cases on my end as well. Just looks like there are some extra files here and there, which I've marked.

**Review by johnjasa — APPROVED** [review]

**Pull request merged**